docs/proposals/003-model-server-protocol/protocol.md

```diff
@@ -0,0 +1,87 @@
+# Endpoint Picker Protocol
+
+The Endpoint Picker, or EPP, is a core component of the inference extension. Ultimately it's
+responsible for picking an endpoint from the `InferencePool`. A reference implementation can be
+found [here](../../../pkg/ext-proc/).
+
+## Proxy Protocol
+
+This is the protocol between the EPP and the proxy (e.g, Envoy).
+
+The EPP MUST implement the Envoy
+[external processing service](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor)protocol.
+
+For each HTTP request, the EPP MUST communicate to the proxy the picked model server endpoint, via
+adding the `target-pod` HTTP header in the request, or otherwise return an error.
+
+## Model Server Protocol
+
+This is the protocol between the EPP and the model servers.
+
+### Inference API Protocol
+
+The model server MUST implement OpenAI’s [Completions](https://platform.openai.com/docs/api-reference/completions)
+and [Chat](https://platform.openai.com/docs/api-reference/chat) APIs.
+
+### Metrics Reporting
+
+The inference extension scrapes metrics from the model servers to make optimal request scheduling
+decisions. The model servers MUST provide the following metrics via a Prometheus endpoint. The exact
+metric names don't necessarily need to be the same as the recommended names here, however the
+metric types and semantics MUST follow this doc.
+
+Note the requirements here are aligned with the
+[model server metrics standardization](https://docs.google.com/document/d/1SpSp1E6moa4HSrJnS4x3NpLuj88sMXr2tbofKlzTZpk)
+effort.
+
+The corresponding metrics in vLLM are also shown in the table below, as vLLM is already integrated
+into the reference endpoint picker implementation.
+
+| Metric | Type | Description | vLLM metric |
+| ----- | ---- | ---- | ---- |
+| TotalQueuedRequests         | Gauge     | The current total number of requests in the queue.| `vllm:num_requests_waiting`|
+| KVCacheUtilization| Gauge     | The current KV cache utilization in percentage.| `vllm:gpu_cache_usage_perc`|
+
+
+### LoRA Adapter Serving
+
+Model servers that support dynamic LoRA serving can benefit from the LoRA affinity algorithm. Note
+the current algorithm in the reference EPP is highly biased towards vLLM's current dynamic LoRA 
+implementation.
+
+The model servers MUST support serving a LoRA adapter specified in the `model` argument of the
+request, provided the requested adapter is valid.
+
+The model server MUST expose the following LoRA adapter information via a RESTful API with response
+in JSON :
```

Comment:
Does the current EPP implementation support this?

Reply:
No it does not. It uses the current vLLM metrics implementation

Reply:
The protocol is part of the release, and so i think we should remove this until the next release when we actually implement it and only spec the metrics based approach.

Reply:
I can document the current metrics approach for vllm, but I don't want to make it a "protocol", because it's really awkward. It's not something we should recommend for the next model server to implement.

The purpose of the protocol is to set up a contract for any new model server integration to follow. So I think we should document this here, and with a note that the current vllm workaround should converge too.

Reply:
But that is why we have versioning, it doesn't make sense to document a contract that the EPP doesn't implement. Once we have EPP implements it, we update the protocol and create a new release. @robscott @smarterclayton for opinions.

Reply:
IMO this is just a short-term workaround and we don't need a protocol to document it - the code speaks for itself. I think the protocol should be something reasonably stable based on our best knowledge, and we are willing to take to integrate with another model server. 

One option is to just not document this at all, until we have some implementation to support it.

Reply:
Not documenting anything is an option, but documenting a protocol and make it part of a release that doesn't implement it is I think not what we should do here. Remember that we are versioning the protocol with the EPP image.

I would still lean towards documenting what we implemented and released, and we can change that in the next release, that is what versioning allows us to do. It is expected that the initial iterations will include more frequent changes, and the protocol will stabilize after that.

Reply:
I would probably align to abdullah here - the goal is to spec what is currently required, so that you don't read things that don't exist.  There will be people using 0.1 for quite a while, so be accurate to 0.1.  Anything that is to be removed can simply be moved to a separate PR which is "draft for 0.2" and not lost.

Reply:
Thanks for the feedback, I updated the doc